### ISSUE TITLE
Let's be really sure about the four Characters.

### ISSUE THREAD

**Issue by jMyles** [nucypher/nucypher-kms#218]
As I have written the logic for the encapsulator ("Enrique"), I have again been wishing to call this Character "Alice" and to move the delegation practices to a yet-unnamed character (I propose one whose name starts with "D" - perhaps "Dennis" - be reminded that "Dan" or "Dave" are sometimes used as generic participants, and so aren't fit names).

Here's the current proposed workforce:

* Alice the Delegator
* Bob the Recipient
* Enrique the Encapsulator / Messenger
* Ursula the re-encryption Proxy

...and here's what I propose as an alternative:

* Alice the Encapsulator / Messenger
* Bob the Recipient
* Dennis the Delegator
* Ursula the re-encryption Proxy

This has the advantage of preserving Alice's role as the owner of the data streams and keeping the narrative wherein Alice has a secret she wants to share with Bob.

It also ensures that the "new" things about PRE (delegation and re-encryption) are the ones with new Character names.

Who's with me?

**Comment by cygnusv** [comment]
I see your point, but I'd prefer to keep it as it currently is, mainly for two reasons:
1) Alice (the delegator) and Bob (the recipient) are the name of the characters in PRE literature. Note that in PRE the emphasis is not in data streams, but in delegation. I acknowledge that we don't have to maintain this tradition, but I don't see a huge (or even small) advantage of changing it.

2) In your proposed approach, I don't see why Alice (the encapsulator) still retains the "role of owner of the data stream". She's the originator, yes, but she is not the owner in the sense that she doesn't control the flow of the data stream; that's controlled solely by the delegator (in your proposed convention, Dennis)

Anyway, I'm not going to take a fundamentalist view on this :) Let's see what the others have to say.

**Comment by jMyles** [comment]
> Alice (the delegator) and Bob (the recipient) are the name of the characters in PRE literature.

But in most cryptological literature, Alice and Bob are the ones exchanging messages, regardless of the power dynamic.

> In your proposed approach, I don't see why Alice (the encapsulator) still retains the "role of owner of the data stream".

Well, how about we consider her at least the owner of the nozzle out of which data pours.  The encapsulator is the owner of heart monitor, the log emission, the fount of secrecy.  Even upon revocation, the encapsulator still has access; to me, that confers a sense of ownership.

**Comment by cygnusv** [comment]
> But in most cryptological literature, Alice and Bob are the ones exchanging messages, regardless of the power dynamic.

That's true, but in this case, Alice doesn't know she's sending a message to Bob. The only thing she knows is that she is sending it initially to Dennis, who can be delegating the message to some Bob (or perhaps not, who knows). So there is not a clear indication from Alice's perspective that there's a Bob. In other words, the whole Alice and Bob dynamic from most cryptological literature cannot be adapted clearly here.

**Comment by jMyles** [comment]
Right.  OK.

Let me sketch for a bit.

**Comment by jMyles** [comment]
OK, I've made quite a lot of progress in the last 9 hours.

After a lot of discussion with @cygnusv, @tuxxy, @michwill, and @mswilkison, the current model I like is no longer the workforce with "Dennis" - or any other 4-character workforce.

Instead, I've created a new model, `DataSource`, which, while having some of the same features as a `Character` (like a `SignatureStamp`) does not have any of the facilities to express agency or ownership.

I think this is a good way to reason about this, because it prevents Alice's devices from being the *owner* of her data, while also allowing them to be discrete `DataSources`, even if she wants to go offline never to return.

So, instead of creating `Enrique`, I have created `nkms.data_sources.DataSource`.  

I'll close this now - for more informaiton, See #217.
